// chai-jest-snapshot/dist/buildConfigState.js
"use strict";

var _buildGetNameForSnapshotUsingTemplate = require("./buildGetNameForSnapshotUsingTemplate");

var _buildGetNameForSnapshotUsingTemplate2 = _interopRequireDefault(_buildGetNameForSnapshotUsingTemplate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = function buildConfigState(determineConfig) {
  var config = {
    snapshotFilename: void 0,
    snapshotNameTemplate: void 0
  };

  function setFilename(snapshotFilename) {
    config.snapshotFilename = snapshotFilename;
  }

  function setTestName(snapshotNameTemplate) {
    config.snapshotNameTemplate = snapshotNameTemplate;
  }

  function configureUsingMochaContext(mochaContext) {
    var currentTest = mochaContext.currentTest;

    setFilename(currentTest.file + ".snap");
    setTestName(currentTest.fullTitle());
  }

  var snapshotNameRegistry = {}; // snapshotNameRegistry[filename][name] => number
  var getNameForSnapshotUsingTemplate = (0, _buildGetNameForSnapshotUsingTemplate2.default)(snapshotNameRegistry);

  function parseArgs(args) {
    return determineConfig(args, config, getNameForSnapshotUsingTemplate);
  }

  function resetSnapshotRegistry() {
    for (var filename in snapshotNameRegistry) {
      delete snapshotNameRegistry[filename];
    }
  }

  return {
    setFilename: setFilename,
    setTestName: setTestName,
    configureUsingMochaContext: configureUsingMochaContext,
    parseArgs: parseArgs,
    resetSnapshotRegistry: resetSnapshotRegistry
  };
};
